feat (FeatureCollection): Add `IFeatureCollection.GetRequiredFeature` extension method

ShreyasJejurkar opened the pull request:
1. Add implementation
2. Add tests

Fixes https://github.com/dotnet/aspnetcore/issues/25471

ShreyasJejurkar committed:
feat (FeatureCollection): Add `IFeatureCollection.GetRequiredFeature` extension method.
1. Add implementation
2. Add tests

ShreyasJejurkar commented:
CI failure are unrelated to this!

davidfowl commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        return (TFeature)featureCollection.GetRequiredFeature(typeof(TFeature));
```
This should be using the generic overload and not fall back to the indexer.

ShreyasJejurkar committed:
Address PR feedback
1. Used generic method instead of indexing!

davidfowl commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,66 @@
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        var feature = featureCollection.Get<TFeature>();
+
+        if (feature == null)
+        {
+            throw new InvalidOperationException($"{nameof(TFeature)} is not available");
+        }
+
+        return feature;
```
```suggestion
        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"{nameof(TFeature)} is not available");
```

ShreyasJejurkar committed:
One-liner because why not!?

davidfowl commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+            throw new ArgumentNullException(nameof(featureType));
+        }
+
+        var feature = featureCollection[featureType];
+
+        if (feature == null)
+        {
+            throw new InvalidOperationException($"{nameof(featureType)} is not available");
+        }
+
+        return feature;
```
Same here?

```suggestion
        return featureCollection[featureType] ?? throw new InvalidOperationException($"{nameof(featureType)} is not available");
```

ShreyasJejurkar replied:
Yess. Changed with the latest commit. 

I am not very much a fan of this one-liner, it does not add any improvement to the resulting codegen, but adds less readability to the code itself! But I am totally ok if it's practice over the repo! :)

ShreyasJejurkar committed:
more one liner!

ShreyasJejurkar commented:
CI failures are unrelated to this. Given file was not edited as part of the commit!

Tratcher commented:
Can you find a few places to use this new API in existing middleware?

ShreyasJejurkar commented:
@Tratcher I will find the spaces where we can use this extension and will use it there with the next commit!

ShreyasJejurkar commented:
Hi @Tratcher, I search through the references of the `Get<TFeature>` method in the codebase for half of an hour, but am not getting the "aaahaaa" moment so that we can use this extension method directly.

In most places if the feature is null we do one of the following - 
1. we are just avoiding certain initialization. literally a guarding `if` condition. 
2. Sometimes we add that feature in FeatureCollection. 
3. uses null coalescing `?.methodCall()` or `!.methodCall()` and forget. 
4. For some places we do throw an `InvalidOperationException` but the problem is the exception message is a different place to place, so if we use our new extension method, then everywhere the same exception message will be thrown and it will be a breaking change for users! I know it's just an exception message change, but still, (I may be wrong) it's breaking change I guess! 

Can you point out any place where you would have thought that we can use this extension method!?

Tratcher commented:
> 4\. For some places we do throw an `InvalidOperationException` but the problem is the exception message is a different place to place, so if we use our new extension method, then everywhere the same exception message will be thrown and it will be a breaking change for users! I know it's just an exception message change, but still, (I may be wrong) it's breaking change I guess!

Changing an exception message is not considered a breaking change, only changing the exception type would be breaking.

Tratcher commented:
> !.methodCall()

These calls should be replaced by this extension. We want to avoid using the `!` null override operator in most code.

pranavkm commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
```
```suggestion
```

ShreyasJejurkar replied:
I tried but got errors as the project targets .NET Framework as well!

pranavkm commented on src/Extensions/Features/test/FeatureCollectionExtensionsTests.cs:
```diff
@@ -0,0 +1,38 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Xunit;
```
```suggestion
```

ShreyasJejurkar replied:
same here as above!

pranavkm commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+    /// </summary>
+    /// <param name="featureCollection"></param>
+    /// <returns>Feature object type</returns>
+    /// <exception cref="InvalidOperationException">Thrown when asked for unregistered feature type</exception>
```
We don't document exceptions elsewhere. I would err on the side of consistency and skip this.

Tratcher replied:
At least mention the exception in the summary.

pranavkm commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+    public static TFeature GetRequiredFeature<TFeature>(this IFeatureCollection featureCollection)
+        where TFeature : notnull
+    {
+        if (featureCollection is null)
+        {
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"{nameof(TFeature)} is not available");
```
```suggestion
        return GetRequiredFeature(featureCollection, typeof(TFeature));
```

davidfowl replied:
No we should be using the generic overload. It's optimized.

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore into feature-ext

ShreyasJejurkar commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+    public static TFeature GetRequiredFeature<TFeature>(this IFeatureCollection featureCollection)
+        where TFeature : notnull
+    {
+        if (featureCollection is null)
+        {
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"{nameof(TFeature)} is not available");
```
Which one? I only see `Get<TFeature>` and the indexer!

ShreyasJejurkar replied:
Hi @pranavkm, before we merge, can you please check this @davidfowl review comment! Do I need to change something here before we merge!?

pranavkm replied:
That was in response to my suggestion which has since been deleted. This can be resolved.

ShreyasJejurkar committed:
fix : xml comments

ShreyasJejurkar committed:
fix - used new extension method instead of using null override operator

ShreyasJejurkar committed:
fix : tests

1. Fix tests exception message
2. Change exception message

pranavkm commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,51 @@
+public static class FeatureCollectionExtensions
+{
+    /// <summary>
+    /// Get feature of type <typeparamref name="TFeature"/> from the <see cref="IFeatureCollection"/>.
+    /// Exception of type <see cref="InvalidOperationException"/> thrown when asked for unregistered feature type
```
```suggestion
    /// Retrieves the requested feature from the collection.
    /// Throws an <see cref="InvalidOperationException"/> if the feature is not present.
```

Using language similar to https://github.com/dotnet/aspnetcore/blob/main/src/Extensions/Features/src/IFeatureCollection.cs#L31-L36

pranavkm commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /// <summary>
+    /// Get feature object of provided type from the <see cref="IFeatureCollection"/>.
+    /// Exception of type <see cref="InvalidOperationException"/> thrown when asked for unregistered feature type
```
```suggestion
    /// Retrieves the requested feature from the collection.
    /// Throws an <see cref="InvalidOperationException"/> if the feature is not present.
```

pranavkm reviewed (commented):
This looks great

ShreyasJejurkar committed:
Refactor : Addressed PR feedback.

1. Updated parameter name in code as as PublicAPI files
2. Removed entry from resources as we no longer need it!
3. Fix corresponding tests.

ShreyasJejurkar commented:
Thanks, @pranavkm for the PR review, appreciates it a lot. 
Addressed them with the latest commit!

pranavkm reviewed (commented):
Looks good to me. @Tratcher / @halter73 could one of you look at the server changes to make sure it's all good?

Tratcher commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"Feature with key {typeof(TFeature)} is not present.");
```
```suggestion
        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"Feature {typeof(TFeature)} is not present.");
```

Tratcher commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+            throw new ArgumentNullException(nameof(key));
+        }
+
+        return featureCollection[key] ?? throw new InvalidOperationException($"{key} is not present.");
```
```suggestion
        return featureCollection[key] ?? throw new InvalidOperationException($"Feature {key} is not present.");
```

ShreyasJejurkar committed:
Update src/Extensions/Features/src/FeatureCollectionExtensions.cs

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

ShreyasJejurkar committed:
Update src/Extensions/Features/src/FeatureCollectionExtensions.cs

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

ShreyasJejurkar committed:
fix : Change exception message in tests.

Tratcher commented:
There are some related test failures in Http3HttpProtocolFeatureCollectionTests

https://dev.azure.com/dnceng/public/_build/results?buildId=1540523&view=ms.vss-test-web.build-test-results-tab&runId=43579686&resultId=120892&paneView=debug

Http3StreamFeatureCollectionDoesIncludeIHttpMinRequestBodyDataRateFeature
System.InvalidOperationException : Feature Microsoft.AspNetCore.Connections.Features.IProtocolErrorCodeFeature is not present.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs:
```diff
@@ -85,9 +86,9 @@ public void Initialize(Http3StreamContext context)
-        _errorCodeFeature = _context.ConnectionFeatures.Get<IProtocolErrorCodeFeature>()!;
-        _streamIdFeature = _context.ConnectionFeatures.Get<IStreamIdFeature>()!;
-        _streamAbortFeature = _context.ConnectionFeatures.Get<IStreamAbortFeature>()!;
+        _errorCodeFeature = _context.ConnectionFeatures.GetRequiredFeature<IProtocolErrorCodeFeature>();
```
You found a bug, this `!` was wrong! Now we get to figure out why. I'll check where IProtocolErrorCodeFeature should be coming from and see why we thought it should have been there.

Tratcher replied:
Oh, its probably the test not setting up all the fields.
https://github.com/dotnet/aspnetcore/blob/4656f159492b1fb021992f52caf82e5bb5a66ee7/src/Servers/Kestrel/shared/test/TestContextFactory.cs#L189
Want to try filling those in with fake features, or do you need some help with that?

ShreyasJejurkar replied:
Yes, @Tratcher I guess. Seems like recent tests failure are related to this only!

halter73 commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+            throw new ArgumentNullException(nameof(featureCollection));
+        }
+
+        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"Feature {typeof(TFeature)} is not present.");
```
```suggestion
        return featureCollection.Get<TFeature>() ?? throw new InvalidOperationException($"Feature '{typeof(TFeature)}' is not present.");
```

davidfowl replied:
Why single quotes?

halter73 replied:
It's something we almost always do when we include parameterized type names in exception messages to make it more obvious that it's a generic error that could happen with different types. We do it all over DI including `GetRequiredService`:

https://github.com/dotnet/runtime/blob/64f48857a6193a81faa32825cc78d7340a3ebd46/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Resources/Strings.resx#L137

We also single quoted type names in some of the error messages this PR is replacing:

https://github.com/dotnet/aspnetcore/pull/39192/files#diff-abeea620ed1d2f72f8243f6600f3dcae70403d3186190f5131b4f909b9173b2fL371

pranavkm replied:
@halter73 could we get this PR merged and do a clean up pass?

Tratcher replied:
There are still failing tests to address, so cleanup is harmless.

halter73 commented on src/Extensions/Features/src/FeatureCollectionExtensions.cs:
```diff
@@ -0,0 +1,52 @@
+            throw new ArgumentNullException(nameof(key));
+        }
+
+        return featureCollection[key] ?? throw new InvalidOperationException($"Feature {key} is not present.");
```
```suggestion
        return featureCollection[key] ?? throw new InvalidOperationException($"Feature '{key}' is not present.");
```

halter73 commented on src/Mvc/Mvc.Core/test/Filters/ResponseCacheFilterExecutorTest.cs:
```diff
@@ -416,7 +416,7 @@ public void NonEmptyVaryByQueryKeys_WithoutConfiguringMiddleware_Throws()
         // Act & Assert
         var exception = Assert.Throws<InvalidOperationException>(() => executor.Execute(context));
-        Assert.Equal("'VaryByQueryKeys' requires the response cache middleware.", exception.Message);
+        Assert.Equal($"Feature {typeof(IResponseCachingFeature)} is not present.", exception.Message);
```
Let's keep the old message for this one since it seems more informative.

ShreyasJejurkar committed:
Addressed PR feedback.

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore into feature-ext

ShreyasJejurkar committed:
fix tests

pranavkm commented:
@ShreyasJejurkar I think the repo this PR was created from was also deleted.

ShreyasJejurkar commented:
Yeah @pranavkm sadly! I lost all of my PR because of I deleted the fork. I read online that if we delete fork, then PR are still merge-able upstream, but seems it's not working!
I think I will close this PR now as it seems like there is no way to merge this! 😑😑

Sorry for inconvenience, and thanks all for your time for reviewing.

davidfowl commented:
😢 We want this feature

ShreyasJejurkar commented:
Yeah, @davidfowl. @pranavkm has open another PR to get this merged https://github.com/dotnet/aspnetcore/pull/39696

Recently I deleted the aspnetcore fork for fresh start, so this thing happened! Sorry for the delay 😔!
